controllers/internal/status/components.go

```diff
@@ -0,0 +1,96 @@
+package status
+
+import (
+	"context"
+
+	"github.com/ViaQ/logerr/kverrors"
+	lokiv1beta1 "github.com/ViaQ/loki-operator/api/v1beta1"
+
+	"github.com/ViaQ/loki-operator/internal/external/k8s"
+	"github.com/ViaQ/loki-operator/internal/manifests"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SetComponentsStatus updates the pod status map component
+func SetComponentsStatus(ctx context.Context, k k8s.Client, req ctrl.Request) error {
+	var s lokiv1beta1.LokiStack
+	if err := k.Get(ctx, req.NamespacedName, &s); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return kverrors.Wrap(err, "failed to lookup lokistack", "name", req.NamespacedName)
+	}
+
+	var (
+		compactorName     = manifests.CompactorName(s.Name)
+		distributorName   = manifests.DistributorName(s.Name)
+		ingesterName      = manifests.IngesterName(s.Name)
+		querierName       = manifests.QuerierName(s.Name)
+		queryFrontendName = manifests.QueryFrontendName(s.Name)
+		psm               = lokiv1beta1.PodStatusMap{}
+	)
+	type componentMapping struct {
+		label string
+		psm   lokiv1beta1.PodStatusMap
+	}
+	components := map[string]componentMapping{
+		compactorName: {
+			label: manifests.LabelCompactorComponent,
+			psm:   psm.DeepCopy(),
+		},
+		distributorName: {
+			label: manifests.LabelDistributorComponent,
+			psm:   psm.DeepCopy(),
+		},
+		ingesterName: {
+			label: manifests.LabelIngesterComponent,
+			psm:   psm.DeepCopy(),
+		},
+		querierName: {
+			label: manifests.LabelQuerierComponent,
+			psm:   psm.DeepCopy(),
+		},
+		queryFrontendName: {
+			label: manifests.LabelQueryFrontendComponent,
+			psm:   psm.DeepCopy(),
+		},
+	}
+
+	for _, c := range components {
+		err := appendPodStatus(ctx, k, c.psm, c.label, s.Name, s.Namespace)
+		if err != nil {
+			return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", c.label)
+		}
+	}
+
+	status := s.Status.DeepCopy()
+	status.Components = lokiv1beta1.LokiStackComponentStatus{
+		Compactor:     components[compactorName].psm,
+		Distributor:   components[distributorName].psm,
+		Ingester:      components[ingesterName].psm,
+		Querier:       components[querierName].psm,
+		QueryFrontend: components[queryFrontendName].psm,
+	}
+	s.Status = *status
+	return k.Status().Update(ctx, &s, &client.UpdateOptions{})
+}
```

Comment:
This whole function is far too complex to avoid a little repetition. It's confusing what is actually happening when trying to read through it. You can remove the custom type structs, pointer copies, indirection, etc, by using a _little_ repetition. 


```// SetComponentsStatus updates the pod status map component
func SetComponentsStatus(ctx context.Context, k k8s.Client, req ctrl.Request) error {
	var s lokiv1beta1.LokiStack
	if err := k.Get(ctx, req.NamespacedName, &s); err != nil {
		if apierrors.IsNotFound(err) {
			return nil
		}
		return kverrors.Wrap(err, "failed to lookup lokistack", "name", req.NamespacedName)
	}


	var err error
	s.Status.Components = lokiv1beta1.LokiStackComponentStatus{}
	if s.Status.Components.Compactor, err = appendPodStatus(ctx, k, manifests.CompactorName(s.Name), s.Name, s.Namespace); err != nil {
		return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", manifests.LabelCompactorComponent)
	}
	if s.Status.Components.Querier, err = appendPodStatus(ctx, k, manifests.QuerierName(s.Name), s.Name, s.Namespace); err != nil {
		return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", manifests.LabelQuerierComponent)
	}
	if s.Status.Components.Distributor, err = appendPodStatus(ctx, k, manifests.DistributorName(s.Name), s.Name, s.Namespace); err != nil {
		return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", manifests.LabelDistributorComponent)
	}
	if s.Status.Components.QueryFrontend, err = appendPodStatus(ctx, k, manifests.QueryFrontendName(s.Name), s.Name, s.Namespace); err != nil {
		return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", manifests.LabelQueryFrontendComponent)
	}
	if s.Status.Components.Ingester, err = appendPodStatus(ctx, k, manifests.IngesterName(s.Name), s.Name, s.Namespace); err != nil {
		return kverrors.Wrap(err, "failed lookup LokiStack component pods status", "name", manifests.LabelIngesterComponent)
	}

	return k.Status().Update(ctx, &s, &client.UpdateOptions{})
}
```

This is a simpler implementation that I think is easier to read and maintain.

Reply:
You can probably shrink it and keep the loop if you make status.Components into a `map[string][]string`. Either way I think it should be rewritten.

Reply:
either ^ or change `status.Components` into a map. no need to swap between the two structures